rotkehlchen/premium/premium.py

```diff
@@ -207,6 +212,78 @@ def set_credentials(self, credentials: PremiumCredentials) -> None:
             self.reset_credentials(old_credentials)
             raise PremiumAuthenticationError('rotki API key was rejected by server')
 
+    def get_remote_devices_information(self) -> dict:
+        """Get the list of devices for the current user"""
+        method = 'manage/premium/devices'
+        signature, data = self.sign(
+            method=method,
+            api_endpoint='webapi',
+        )
+        self.session.headers.update({'API-SIGN': base64.b64encode(signature.digest())})
+
+        try:
+            response = self.session.get(
+                f'{self.rotki_web}{method}',
+                data=data,
+                timeout=ROTKEHLCHEN_SERVER_TIMEOUT,
+            )
+        except requests.exceptions.RequestException as e:
+            msg = f'Could not connect to rotki server due to {e!s}'
+            log.error(msg)
+            raise RemoteError(msg) from e
+
+        data = _process_dict_response(response)
+        return data
+
+    def authenticate_device(self) -> None:
+        """
+        Check if the device is in the list of the devices and if it isn't add it when possible.
+        May raise:
+        - PremiumAuthenticationError: when the device can't be registered
+        """
+        device_data = self.get_remote_devices_information()
+        num_devices = len(device_data['devices'])
+        devices_limit = device_data['limit']
+        device_id = machineid.hashed_id(self.username)
+
+        device_found = False
+        for device in device_data['devices']:
+            if device['device_identifier'] != device_id:
+                continue
+            device_found = True
+            break
+
+        if not device_found:
+            if num_devices < devices_limit:
+                # try to sign up the device
+                self._register_new_device(device_id)
+            else:
+                # user has to edit his devices
+                raise PremiumAuthenticationError('The limit of devices has been reached')
+
+    def _register_new_device(self, device_id: str) -> dict:
+        log.debug(f'Registering new device {device_id}')
+        method = 'devices'
+        device_name = platform.system()
+        signature, data = self.sign(
+            method=method,
+            device_identifier=device_id,
+            device_name=device_name,
+        )
+        self.session.headers.update({'API-SIGN': base64.b64encode(signature.digest())})
```

Comment:
This signature creation is both here and in the previous function. Abstract it and call it in both places

Reply:
Moved it inside the sign method and removed the header set from where it is called